sympy/algebras/euler.py

```diff
@@ -0,0 +1,634 @@
+
+from mpmath.libmp.libmpf import prec_to_dps
+
+
+def _check_sequence(seq):
+    """validate seq and return info"""
+    if type(seq) != str:
+        raise ValueError('Expected seq to be a string.')
+    if len(seq) != 3:
+        raise ValueError("Expected 3 axes, got `{}`.".format(seq))
+
+    intrinsic = seq.isupper()
+    extrinsic = seq.islower()
+    if not (intrinsic or extrinsic):
+        raise ValueError("seq must either be fully uppercase (for extrinsic "
+                         "rotations), or fully lowercase, for intrinsic "
+                         "rotations).")
+
+    i, j, k = seq.lower()
+    if (i == j) or (j == k):
+        raise ValueError("Consecutive axes must be different")
+
+    bad = set(seq) - set('xyzXYZ')
+    if bad:
+        raise ValueError("Expected axes from `seq` to be from "
+                         "['x', 'y', 'z'] or ['X', 'Y', 'Z'], "
+                         "got {}".format(''.join(bad)))
+
+    # get elementary basis vectors
+    ei = [1 if n == i else 0 for n in 'xyz']
+    ej = [1 if n == j else 0 for n in 'xyz']
+    ek = [1 if n == k else 0 for n in 'xyz']
+
+    # get indices
+    i = 'xyz'.index(i) + 1
+    j = 'xyz'.index(j) + 1
+    k = 'xyz'.index(k) + 1
+
+    info = {
+            'seq': seq,
+            'extrinsic': extrinsic,
+            'intrinsic': intrinsic,
+            'symmetric': i == k,
+            'i': i,
+            'j': j,
+            'k': k,
+            'ei': ei,
+            'ej': ej,
+            'ek': ek
+            }
+
+    return info
+
+
+def _wrap(angle):
+    if angle.is_number:
+        return (angle + pi) % (2 * pi) - pi
+    else:
+        return angle
+
+
+class Euler(Expr):
+    """Provides Euler angles operations.
+
+    Euler objects can be instantiated as Euler(alpha, beta, gamma, seq).
+
+    Parameters
+    ==========
+
+    seq : string of length 3
+        Represents the sequence of rotations.
+        For intrinsic rotations, seq must be all lowercase and its elements
+        must be from the set `{'x', 'y', 'z'}`
+        For extrinsic rotations, seq must be all uppercase and its elements
+        must be from the set `{'X', 'Y', 'Z'}`
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Euler_angles
+
+    """
+    _op_priority = 11.0
+
+    is_commutative = False
+
+    def __new__(cls, alpha=0, beta=0, gamma=0, seq='ZYX'):
+        info = _check_sequence(seq)
+        alpha, beta, gamma = map(sympify, (alpha, beta, gamma))
+        alpha, beta, gamma = map(_wrap, (alpha, beta, gamma))
+
+        if any(i.is_commutative is False for i in [alpha, beta, gamma]):
+            raise ValueError("arguments have to be commutative")
+        else:
+            obj = Expr.__new__(cls, alpha, beta, gamma)
+            obj._alpha = alpha
+            obj._beta = beta
+            obj._gamma = gamma
+            obj._info = info
+            obj._q = None
+            obj._R = None
+            return obj
```

Comment:
Basic subclasses should avoid creating lots of attributes that are not part of the args of the object.

This will not satisfy the func-args invariant: `obj == obj.func(*obj.args)` whenever `seq` is something other than `'ZYX'`.

Reply:
Added `seq` to args, but now for some reason I have a bug with `evalf`:
```python
    euler2.evalf()

  File /sympy/sympy/core/evalf.py:1644 in evalf
    result = evalf(self, prec + 4, options)

  File /sympy/sympy/core/evalf.py:1492 in evalf
    re, im = as_real_imag()

  File /sympy/sympy/core/expr.py:1931 in as_real_imag
    return (re(self), im(self))

  File /sympy/sympy/core/cache.py:72 in wrapper
    retval = cfunc(*args, **kwargs)

  File /sympy/sympy/core/function.py:470 in __new__
    result = super().__new__(cls, *args, **options)

  File /sympy/sympy/core/cache.py:72 in wrapper
    retval = cfunc(*args, **kwargs)

  File /sympy/sympy/core/function.py:310 in __new__
    evaluated = cls.eval(*args)

  File sympy/sympy/functions/elementary/complexes.py:92 in eval
    elif not term.has(I) and term.is_extended_real:

  File sympy/sympy/core/cache.py:72 in wrapper
    retval = cfunc(*args, **kwargs)

  File sympy/sympy/core/basic.py:1267 in has
    return self._has(iterargs, *patterns)

  File sympy/sympy/core/basic.py:1357 in _has
    for i in iterargs(self):  #

  File /sympy/sympy/core/traversal.py:31 in iterargs
    args.extend(i.args)

AttributeError: 'str' object has no attribute 'args'
```

I think I should sympify `seq`, but then I'd need to recover the string from the symbol. How can I do that?

Reply:
A basic workaround is to use `core.Symbol`, or subclass nullary `Basic` instance of each alphabets
(`class X(Basic): pass`, ...)

Reply:
Alright I think I managed to work around this!